virtualizarr/parsers/zarr.py

```diff
@@ -26,64 +26,272 @@
+                array_name = zarr_array.name.lstrip("/") if zarr_array.name else "array"
+                v3_dict["dimension_names"] = [
+                    f"{array_name}_dim_{i}" for i in range(len(zarr_array.shape))
+                ]
+            v3_metadata = ArrayV3Metadata.from_dict(v3_dict)
+
+        # Replace V2ChunkKeyEncoding with V3 DefaultChunkKeyEncoding
+        # The automatic conversion preserves V2's encoding, causing zarr to use V2-style
+        # paths (array/0) instead of V3-style (array/c/0). This ensures V3 semantics.
+        v3_dict = v3_metadata.to_dict()
+        v3_dict["chunk_key_encoding"] = {"name": "default", "separator": "."}
+        v3_metadata = ArrayV3Metadata.from_dict(v3_dict)
+
+        return v3_metadata
+
+
+class ZarrV3Strategy:
+    """Strategy for handling Zarr V3 arrays."""
+
+    async def get_chunk_mapping(
+        self, zarr_array: ZarrArrayType, path: str
+    ) -> dict[str, dict[str, Any]]:
+        """
+        Create a mapping of chunk coordinates to their storage locations for V3 arrays.
+
+        V3 uses paths like: array_name/c/0/1/2
+        """
+        name = getattr(zarr_array, "name", "") or ""
+        name = name.lstrip("/")
+
+        # Handle scalar arrays
+        if zarr_array.shape == ():
+            prefix = f"{name}/c" if name else "c"
+            size = await zarr_array.store.getsize(prefix)
+            return {"c": {"path": join_url(path, prefix), "offset": 0, "length": size}}
+
+        prefix = f"{name}/c/" if name else "c/"
         prefix_keys = [(x,) async for x in zarr_array.store.list_prefix(prefix)]
+        if not prefix_keys:
+            return {}
+
         _lengths = await _concurrent_map(prefix_keys, zarr_array.store.getsize)
         chunk_keys = [x[0].split(prefix)[1] for x in prefix_keys]
         _dict_keys = [key.replace("/", ".") for key in chunk_keys]
-        _paths = [path + "/" + prefix + key for key in chunk_keys]
-
-    _offsets = [0] * len(_lengths)
-    return {
-        key: {"path": path, "offset": offset, "length": length}
-        for key, path, offset, length in zip(
-            _dict_keys,
-            _paths,
-            _offsets,
-            _lengths,
-        )
-    }
+        _paths = [join_url(path, prefix + key) for key in chunk_keys]
+        _offsets = [0] * len(_lengths)
 
+        return {
+            key: {"path": p, "offset": offset, "length": length}
+            for key, p, offset, length in zip(_dict_keys, _paths, _offsets, _lengths)
+        }
 
-async def build_chunk_manifest(zarr_array: ZarrArrayType, path: str) -> ChunkManifest:
-    """Build a ChunkManifest from a dictionary"""
-    chunk_map = await get_chunk_mapping_prefix(zarr_array=zarr_array, path=path)
-    return ChunkManifest(chunk_map)
+    def get_metadata(self, zarr_array: ZarrArrayType) -> ArrayV3Metadata:
+        """Return V3 metadata as-is (no conversion needed)."""
+        return zarr_array.metadata  # type: ignore[return-value]
```

Comment:
An important thing that we should really already be doing is raising if the zarr V3 array contains [sharding](https://zarr.readthedocs.io/en/stable/user-guide/performance.html#sharding), because we don't yet support that (and if we don't raise then our `offset=0` means we will silently return corrupted results 😬 ).